source/dofs/dof_tools.cc

```diff
@@ -765,6 +765,110 @@ namespace DoFTools
 
 
 
+  template <typename DoFHandlerType>
+  IndexSet
+  extract_dofs_with_support_on(const DoFHandlerType &dof_handler,
+                               const std::function< bool(const typename DoFHandlerType::active_cell_iterator &)> &predicate,
+                               const ConstraintMatrix &cm)
+  {
+    const std::function< bool(const typename DoFHandlerType::active_cell_iterator &)> predicate_local
+    = [=] (const typename DoFHandlerType::active_cell_iterator & cell) -> bool
+    {
+      return cell->is_locally_owned() && predicate(cell);
+    };
+
+    std::vector<types::global_dof_index> local_dof_indices;
+    local_dof_indices.reserve (max_dofs_per_cell(dof_handler));
+
+    // Get all the dofs that live on the subdomain:
+    std::set<types::global_dof_index> predicate_dofs;
+
+    typename DoFHandlerType::active_cell_iterator
+    cell = dof_handler.begin_active(),
+    endc = dof_handler.end();
+    for (; cell!=endc; ++cell)
+      if (!cell->is_artificial() && predicate(cell))
+        {
+          local_dof_indices.resize (cell->get_fe().dofs_per_cell);
+          cell->get_dof_indices (local_dof_indices);
+          predicate_dofs.insert(local_dof_indices.begin(),
+                                local_dof_indices.end());
+        }
+
+    // Get halo layer and accumulate its DoFs
+    std::set<types::global_dof_index> halo_dofs;
```

Comment:
could you rename this variable to `dofs_with_support_on_halo_cells`? I think this makes its meaning clearer, in particular when later on you add the `extra_halo` dofs to this set.